### PR TITLE
[Rust] Basic reading support for nested fields.

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/io/reader.rs [hunk 1]
```diff
@@ -208,31 +211,64 @@ impl<'a> FileReader<'a> {
...
-            println!("Reading: {}", field);
-            let page_info = self.page_table.get(field.id, batch_id).unwrap();
-            Ok(Arc::new(Int32Array::from(
-                (0..page_info.length as i32).collect::<Vec<_>>(),
-            )))
+    async fn read_struct_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        // TODO: use tokio to make the reads in parallel.
+        let mut sub_arrays = vec![];
+        for child in field.children.as_slice() {
+            // let arr = self.read_array(, batch_id)
```
Comment (changhiskhan) [inline rust/src/io/reader.rs:218]: remove?

File: rust/src/io/reader.rs [hunk 2]
```diff
@@ -208,31 +211,64 @@ impl<'a> FileReader<'a> {
...
         }
+
+        Ok(Arc::new(StructArray::from(sub_arrays)))
+    }
+
+    async fn read_list_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        let page_info = self.page_info(field, batch_id)?;
+        let position_arr = self
+            .object_reader
+            .read_primitive_array(&DataType::Int32, page_info.position, page_info.length)
```
Comment (changhiskhan) [inline rust/src/io/reader.rs:230]: i guess we could refactor this later into read_list_array and read_large_list_array that calls a parametrized method with idnex type

File: rust/src/io/reader.rs [hunk 3]
```diff
@@ -208,31 +211,64 @@ impl<'a> FileReader<'a> {
...
+            .await?;
+        let positions = position_arr.as_any().downcast_ref::<Int32Array>().unwrap();
```
Comment (changhiskhan) [inline rust/src/io/reader.rs:232]: does `let positions: &Int32Array = as_primitive_array(position_arr.as_ref())` work?
it's not that much code savings but just hides some of the casting and unwrapping boilerplate.

File: rust/src/io/reader.rs [hunk 4]
```diff
@@ -208,31 +211,64 @@ impl<'a> FileReader<'a> {
...
+    async fn read_struct_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        // TODO: use tokio to make the reads in parallel.
+        let mut sub_arrays = vec![];
+        for child in field.children.as_slice() {
+            // let arr = self.read_array(, batch_id)
+            let arr = self.read_array(&child, batch_id).await?;
+            sub_arrays.push((child.into(), arr));
         }
+
+        Ok(Arc::new(StructArray::from(sub_arrays)))
```
Comment (changhiskhan) [inline rust/src/io/reader.rs:223]: one really annoying thing for performance is that the sub_arrays get cloned inside the factor function here.

**Review by changhiskhan — APPROVED** [review]
couple of questions / nits

File: rust/src/io/reader.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/io/reader.rs:230]: yes, you are right.

File: rust/src/io/reader.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/io/reader.rs:218]: fixed.

File: rust/src/io/reader.rs (hunk 4 above)
Comment (eddyxu) [reply rust/src/io/reader.rs:223]: Arrow should do shallow copy at the buffer level.

File: rust/src/io/reader.rs (hunk 3 above)
Comment (eddyxu) [reply rust/src/io/reader.rs]: TIL

**Pull request merged**